lib/portlayer/attach/communication/connector.go

```diff
@@ -112,33 +128,44 @@ func (c *Connector) Interaction(ctx context.Context, id string) (SessionInteract
 func (c *Connector) interaction(ctx context.Context, id string) (SessionInteractor, error) {
 	defer trace.End(trace.Begin(id))
 
-	conn := c.aliveAndKicking(ctx, id)
-	if conn != nil {
-		return conn, nil
-	}
-
 	if ctx.Err() == context.DeadlineExceeded {
 		return nil, fmt.Errorf("attach connector: no such connection")
 	}
 
+	conn, err := c.SessionIfAlive(ctx, id)
+	if conn != nil && err == nil {
+		return conn, nil
+	}
+
 	result := make(chan SessionInteractor, 1)
 	go func() {
 		ok := false
-		var conn SessionInteractor
+		var v *LazySessionInteractor
 
 		c.mutex.RLock()
 		defer c.mutex.RUnlock()
 
 		for !ok && ctx.Err() == nil {
-			conn, ok = c.interactions[id]
+			v, ok = c.interactions[id]
 			if ok {
-				// no need to test this as we just created it
-				if err := conn.Unblock(); err == nil {
-					log.Infof("attach connector: Unblocked %s, returning", id)
+				conn, err := v.Initialize()
+				if conn != nil && err == nil {
+					// no need to test this connection as we just created it, unblock if needed
+					log.Debugf("attach connector: Unblocking for %s", id)
+					err = conn.Unblock()
```

Comment:
There is a race in the failure/reset path of Unblock vs other Unblock calls - don't know if it's important.